domain/crossmodelrelation/state/model/remoteapplication.go

```diff
@@ -1008,6 +1059,144 @@ func (st *State) addCharmRelations(ctx context.Context, tx *sqlair.TX, uuid stri
 	return nil
 }
 
+func (st *State) getCharmUUIDByApplicationUUID(ctx context.Context, tx *sqlair.TX, appUUID string) (string, error) {
+	input := uuid{UUID: appUUID}
+
+	query := `
+SELECT charm_uuid AS &uuid.uuid
+FROM application
+WHERE uuid = $uuid.uuid AND life_id < 2;
+`
+	queryStmt, err := st.Prepare(query, input)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	var charmUUID uuid
+
+	if err := tx.Query(ctx, queryStmt, input).Get(&charmUUID); errors.Is(err, sqlair.ErrNoRows) {
+		return "", applicationerrors.ApplicationNotFound
+	} else if err != nil {
+		return "", errors.Errorf("querying charm UUID for application %q: %w", appUUID, err)
+	}
+
+	return charmUUID.UUID, nil
+}
+
+func (st *State) insertNetNode(
+	ctx context.Context,
+	tx *sqlair.TX,
+	netNodeUUID string,
+) error {
+	input := uuid{UUID: netNodeUUID}
+
+	query := `INSERT INTO net_node (uuid) VALUES ($uuid.*)`
+	queryStmt, err := st.Prepare(query, input)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, queryStmt, input).Run(); err != nil {
+		return errors.Errorf("creating net node for machine: %w", err)
+	}
+
+	return nil
+}
+
+func (st *State) getMissingApplicationUnits(
+	ctx context.Context,
+	tx *sqlair.TX,
+	appUUID string,
+	units []string,
+) ([]string, error) {
+	if len(units) == 0 {
+		return nil, nil
+	}
+
+	type unitName struct {
+		Name string `db:"name"`
+	}
+	type unitNames []string
+	names := unitNames(units)
+	appUUIDRec := uuid{UUID: appUUID}
+
+	query := `
+SELECT name AS &unitName.name
+FROM   unit
+WHERE  application_uuid = $uuid.uuid
+AND    name IN ($unitNames[:]);
+`
+
+	queryStmt, err := st.Prepare(query, names, unitName{}, appUUIDRec)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var foundUnits []unitName
+	if err := tx.Query(ctx, queryStmt, names, appUUIDRec).GetAll(&foundUnits); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("querying existing units for application %q: %w", appUUID, err)
+	}
+
+	foundUnitSet := make(map[string]struct{})
```

Comment:
suggestion: use set.Strings instead, it has helper methods to do what is done here, including `Difference`.